File: artifacts/heman/refactor/aabb2.c
#include <aabb2.c>
#include <aabb2.h>
#include <math.h>
#include <utility.h>
#include <vec2.h>

extern float kmSQR(float s);
extern float kmDegreesToRadians(float degrees);
extern float kmRadiansToDegrees(float radians);
extern float kmMin(float lhs, float rhs);
extern float kmMax(float lhs, float rhs);
extern unsigned char kmAlmostEqual(float lhs, float rhs);
extern float kmClamp(float x, float min, float max);
extern float kmLerp(float x, float y, float factor);
struct kmMat3;
#pragma pack(push)
#pragma pack(1)
typedef struct kmVec2
{
  float x;
  float y;
} kmVec2;
#pragma pack(pop)
kmVec2 *kmVec2Fill(kmVec2 *pOut, float x, float y);
float kmVec2Length(const kmVec2 *pIn);
float kmVec2LengthSq(const kmVec2 *pIn);
kmVec2 *kmVec2Normalize(kmVec2 *pOut, const kmVec2 *pIn);
kmVec2 *kmVec2Lerp(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2, float t);
kmVec2 *kmVec2Add(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
float kmVec2Dot(const kmVec2 *pV1, const kmVec2 *pV2);
float kmVec2Cross(const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Subtract(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Mul(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Div(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Transform(kmVec2 *pOut, const kmVec2 *pV1, const struct kmMat3 *pM);
kmVec2 *kmVec2TransformCoord(kmVec2 *pOut, const kmVec2 *pV, const struct kmMat3 *pM);
kmVec2 *kmVec2Scale(kmVec2 *pOut, const kmVec2 *pIn, const float s);
int kmVec2AreEqual(const kmVec2 *p1, const kmVec2 *p2);
kmVec2 *kmVec2Assign(kmVec2 *pOut, const kmVec2 *pIn);
kmVec2 *kmVec2RotateBy(kmVec2 *pOut, const kmVec2 *pIn, const float degrees, const kmVec2 *center);
float kmVec2DegreesBetween(const kmVec2 *v1, const kmVec2 *v2);
float kmVec2DistanceBetween(const kmVec2 *v1, const kmVec2 *v2);
kmVec2 *kmVec2MidPointBetween(kmVec2 *pOut, const kmVec2 *v1, const kmVec2 *v2);
kmVec2 *kmVec2Reflect(kmVec2 *pOut, const kmVec2 *pIn, const kmVec2 *normal);
extern const kmVec2 KM_VEC2_POS_Y;
extern const kmVec2 KM_VEC2_NEG_Y;
extern const kmVec2 KM_VEC2_NEG_X;
extern const kmVec2 KM_VEC2_POS_X;
extern const kmVec2 KM_VEC2_ZERO;
typedef struct kmAABB2
{
  kmVec2 min;
  kmVec2 max;
} kmAABB2;
kmAABB2 *kmAABB2Initialize(kmAABB2 *pBox, const kmVec2 *centre, const float width, const float height, const float depth);
kmAABB2 *kmAABB2Sanitize(kmAABB2 *pOut, const kmAABB2 *pIn);
int kmAABB2ContainsPoint(const kmAABB2 *pBox, const kmVec2 *pPoint);
kmAABB2 *kmAABB2Assign(kmAABB2 *pOut, const kmAABB2 *pIn);
kmAABB2 *kmAABB2Translate(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *translation);
kmAABB2 *kmAABB2Scale(kmAABB2 *pOut, const kmAABB2 *pIn, float s);
kmAABB2 *kmAABB2ScaleWithPivot(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *pivot, float s);
unsigned int kmAABB2ContainsAABB(const kmAABB2 *container, const kmAABB2 *to_check);
float kmAABB2DiameterX(const kmAABB2 *aabb);
float kmAABB2DiameterY(const kmAABB2 *aabb);
kmVec2 *kmAABB2Centre(const kmAABB2 *aabb, kmVec2 *pOut);
kmAABB2 *kmAABB2ExpandToContain(kmAABB2 *pOut, const kmAABB2 *pIn, const kmAABB2 *other);
kmAABB2 *kmAABB2Initialize(kmAABB2 *pBox, const kmVec2 *centre, const float width, const float height, const float depth)
{
  if (!pBox)
  {
    return 0;
  }
  kmVec2 origin;
  kmVec2 *point = (centre) ? ((kmVec2 *) centre) : (&origin);
  kmVec2Fill(&origin, .0f, .0f);
  pBox->min.x = point->x - (width / 2);
  pBox->min.y = point->y - (height / 2);
  pBox->max.x = point->x + (width / 2);
  pBox->max.y = point->y + (height / 2);
  return pBox;
}

kmAABB2 *kmAABB2Sanitize(kmAABB2 *pOut, const kmAABB2 *pIn)
{
  if (pIn->min.x <= pIn->max.x)
  {
    pOut->min.x = pIn->min.x;
    pOut->max.x = pIn->max.x;
  }
  else
  {
    pOut->min.x = pIn->max.x;
    pOut->max.x = pIn->min.x;
  }
  if (pIn->min.y <= pIn->max.y)
  {
    pOut->min.y = pIn->min.y;
    pOut->max.y = pIn->max.y;
  }
  else
  {
    pOut->min.y = pIn->max.y;
    pOut->max.y = pIn->min.y;
  }
  return pOut;
}

int kmAABB2ContainsPoint(const kmAABB2 *pBox, const kmVec2 *pPoint)
{
  if ((((pPoint->x >= pBox->min.x) && (pPoint->x <= pBox->max.x)) && (pPoint->y >= pBox->min.y)) && (pPoint->y <= pBox->max.y))
  {
    return 1;
  }
  return 0;
}

kmAABB2 *kmAABB2Assign(kmAABB2 *pOut, const kmAABB2 *pIn)
{
  kmVec2Assign(&pOut->min, &pIn->min);
  kmVec2Assign(&pOut->max, &pIn->max);
  return pOut;
}

kmAABB2 *kmAABB2Translate(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *translation)
{
  kmVec2Add(&pOut->min, &pIn->min, translation);
  kmVec2Add(&pOut->max, &pIn->max, translation);
  return pOut;
}

kmAABB2 *kmAABB2Scale(kmAABB2 *pOut, const kmAABB2 *pIn, float s)
{
  kmVec2Scale(&pOut->max, &pIn->max, s);
  kmVec2Scale(&pOut->min, &pIn->min, s);
  return pOut;
}

kmAABB2 *kmAABB2ScaleWithPivot(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *pivot, float s)
{
  kmVec2 translate;
  translate.x = -pivot->x;
  translate.y = -pivot->y;
  kmAABB2Translate(pOut, pIn, &translate);
  kmAABB2Scale(pOut, pIn, s);
  kmAABB2Translate(pOut, pIn, pivot);
  return pOut;
}

unsigned int kmAABB2ContainsAABB(const kmAABB2 *container, const kmAABB2 *to_check)
{
  kmVec2 corners[4];
  kmVec2Fill(&corners[0], to_check->min.x, to_check->min.y);
  kmVec2Fill(&corners[1], to_check->max.x, to_check->min.y);
  kmVec2Fill(&corners[2], to_check->max.x, to_check->max.y);
  kmVec2Fill(&corners[3], to_check->min.x, to_check->max.y);
  int nContains = ((kmAABB2ContainsPoint(container, &corners[0]) + kmAABB2ContainsPoint(container, &corners[1])) + kmAABB2ContainsPoint(container, &corners[2])) + kmAABB2ContainsPoint(container, &corners[3]);
  if (nContains == 0)
  {
    return (unsigned int) 0;
  }
  else
    if (nContains < 4)
  {
    return (unsigned int) 1;
  }
  else
  {
    return (unsigned int) 2;
  }
}

float kmAABB2DiameterX(const kmAABB2 *aabb)
{
  return aabb->max.x - aabb->min.x;
}

float kmAABB2DiameterY(const kmAABB2 *aabb)
{
  return aabb->max.y - aabb->min.y;
}

kmVec2 *kmAABB2Centre(const kmAABB2 *aabb, kmVec2 *pOut)
{
  kmVec2Add(pOut, &aabb->min, &aabb->max);
  kmVec2Scale(pOut, pOut, 0.5);
  return pOut;
}

kmAABB2 *kmAABB2ExpandToContain(kmAABB2 *pOut, const kmAABB2 *pIn, const kmAABB2 *other)
{
  kmAABB2 result;
  result.min.x = (pIn->min.x < other->min.x) ? (pIn->min.x) : (other->min.x);
  result.max.x = (pIn->max.x > other->max.x) ? (pIn->max.x) : (other->max.x);
  result.min.y = (pIn->min.y < other->min.y) ? (pIn->min.y) : (other->min.y);
  result.max.y = (pIn->max.y > other->max.y) ? (pIn->max.y) : (other->max.y);
  kmAABB2Assign(pOut, &result);
  return pOut;
}


extern float kmSQR(float s);
extern float kmDegreesToRadians(float degrees);
extern float kmRadiansToDegrees(float radians);
extern float kmMin(float lhs, float rhs);
extern float kmMax(float lhs, float rhs);
extern unsigned char kmAlmostEqual(float lhs, float rhs);
extern float kmClamp(float x, float min, float max);
extern float kmLerp(float x, float y, float factor);
struct kmMat3;
#pragma pack(push)
#pragma pack(1)
typedef struct kmVec2
{
  float x;
  float y;
} kmVec2;
#pragma pack(pop)
kmVec2 *kmVec2Fill(kmVec2 *pOut, float x, float y);
float kmVec2Length(const kmVec2 *pIn);
float kmVec2LengthSq(const kmVec2 *pIn);
kmVec2 *kmVec2Normalize(kmVec2 *pOut, const kmVec2 *pIn);
kmVec2 *kmVec2Lerp(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2, float t);
kmVec2 *kmVec2Add(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
float kmVec2Dot(const kmVec2 *pV1, const kmVec2 *pV2);
float kmVec2Cross(const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Subtract(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Mul(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Div(kmVec2 *pOut, const kmVec2 *pV1, const kmVec2 *pV2);
kmVec2 *kmVec2Transform(kmVec2 *pOut, const kmVec2 *pV1, const struct kmMat3 *pM);
kmVec2 *kmVec2TransformCoord(kmVec2 *pOut, const kmVec2 *pV, const struct kmMat3 *pM);
kmVec2 *kmVec2Scale(kmVec2 *pOut, const kmVec2 *pIn, const float s);
int kmVec2AreEqual(const kmVec2 *p1, const kmVec2 *p2);
kmVec2 *kmVec2Assign(kmVec2 *pOut, const kmVec2 *pIn);
kmVec2 *kmVec2RotateBy(kmVec2 *pOut, const kmVec2 *pIn, const float degrees, const kmVec2 *center);
float kmVec2DegreesBetween(const kmVec2 *v1, const kmVec2 *v2);
float kmVec2DistanceBetween(const kmVec2 *v1, const kmVec2 *v2);
kmVec2 *kmVec2MidPointBetween(kmVec2 *pOut, const kmVec2 *v1, const kmVec2 *v2);
kmVec2 *kmVec2Reflect(kmVec2 *pOut, const kmVec2 *pIn, const kmVec2 *normal);
extern const kmVec2 KM_VEC2_POS_Y;
extern const kmVec2 KM_VEC2_NEG_Y;
extern const kmVec2 KM_VEC2_NEG_X;
extern const kmVec2 KM_VEC2_POS_X;
extern const kmVec2 KM_VEC2_ZERO;
typedef struct kmAABB2
{
  kmVec2 min;
  kmVec2 max;
} kmAABB2;
kmAABB2 *kmAABB2Initialize(kmAABB2 *pBox, const kmVec2 *centre, const float width, const float height, const float depth);
kmAABB2 *kmAABB2Sanitize(kmAABB2 *pOut, const kmAABB2 *pIn);
int kmAABB2ContainsPoint(const kmAABB2 *pBox, const kmVec2 *pPoint);
kmAABB2 *kmAABB2Assign(kmAABB2 *pOut, const kmAABB2 *pIn);
kmAABB2 *kmAABB2Translate(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *translation);
kmAABB2 *kmAABB2Scale(kmAABB2 *pOut, const kmAABB2 *pIn, float s);
kmAABB2 *kmAABB2ScaleWithPivot(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *pivot, float s);
unsigned int kmAABB2ContainsAABB(const kmAABB2 *container, const kmAABB2 *to_check);
float kmAABB2DiameterX(const kmAABB2 *aabb);
float kmAABB2DiameterY(const kmAABB2 *aabb);
kmVec2 *kmAABB2Centre(const kmAABB2 *aabb, kmVec2 *pOut);
kmAABB2 *kmAABB2ExpandToContain(kmAABB2 *pOut, const kmAABB2 *pIn, const kmAABB2 *other);
kmAABB2 *kmAABB2Initialize(kmAABB2 *pBox, const kmVec2 *centre, const float width, const float height, const float depth)
{
  if (!pBox)
  {
    return 0;
  }
  kmVec2 origin;
  kmVec2 *point = (centre) ? ((kmVec2 *) centre) : (&origin);
  kmVec2Fill(&origin, .0f, .0f);
  pBox->min.x = point->x - (width / 2);
  pBox->min.y = point->y - (height / 2);
  pBox->max.x = point->x + (width / 2);
  pBox->max.y = point->y + (height / 2);
  return pBox;
}

kmAABB2 *kmAABB2Sanitize(kmAABB2 *pOut, const kmAABB2 *pIn)
{
  if (pIn->min.x <= pIn->max.x)
  {
    pOut->min.x = pIn->min.x;
    pOut->max.x = pIn->max.x;
  }
  else
  {
    pOut->min.x = pIn->max.x;
    pOut->max.x = pIn->min.x;
  }
  if (pIn->min.y <= pIn->max.y)
  {
    pOut->min.y = pIn->min.y;
    pOut->max.y = pIn->max.y;
  }
  else
  {
    pOut->min.y = pIn->max.y;
    pOut->max.y = pIn->min.y;
  }
  return pOut;
}

int kmAABB2ContainsPoint(const kmAABB2 *pBox, const kmVec2 *pPoint)
{
  if ((((pPoint->x >= pBox->min.x) && (pPoint->x <= pBox->max.x)) && (pPoint->y >= pBox->min.y)) && (pPoint->y <= pBox->max.y))
  {
    return 1;
  }
  return 0;
}

kmAABB2 *kmAABB2Assign(kmAABB2 *pOut, const kmAABB2 *pIn)
{
  kmVec2Assign(&pOut->min, &pIn->min);
  kmVec2Assign(&pOut->max, &pIn->max);
  return pOut;
}

kmAABB2 *kmAABB2Translate(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *translation)
{
  kmVec2Add(&pOut->min, &pIn->min, translation);
  kmVec2Add(&pOut->max, &pIn->max, translation);
  return pOut;
}

kmAABB2 *kmAABB2Scale(kmAABB2 *pOut, const kmAABB2 *pIn, float s)
{
  kmVec2Scale(&pOut->max, &pIn->max, s);
  kmVec2Scale(&pOut->min, &pIn->min, s);
  return pOut;
}

kmAABB2 *kmAABB2ScaleWithPivot(kmAABB2 *pOut, const kmAABB2 *pIn, const kmVec2 *pivot, float s)
{
  kmVec2 translate;
  translate.x = -pivot->x;
  translate.y = -pivot->y;
  kmAABB2Translate(pOut, pIn, &translate);
  kmAABB2Scale(pOut, pIn, s);
  kmAABB2Translate(pOut, pIn, pivot);
  return pOut;
}

unsigned int kmAABB2ContainsAABB(const kmAABB2 *container, const kmAABB2 *to_check)
{
  kmVec2 corners[4];
  kmVec2Fill(&corners[0], to_check->min.x, to_check->min.y);
  kmVec2Fill(&corners[1], to_check->max.x, to_check->min.y);
  kmVec2Fill(&corners[2], to_check->max.x, to_check->max.y);
  kmVec2Fill(&corners[3], to_check->min.x, to_check->max.y);
  int nContains = ((kmAABB2ContainsPoint(container, &corners[0]) + kmAABB2ContainsPoint(container, &corners[1])) + kmAABB2ContainsPoint(container, &corners[2])) + kmAABB2ContainsPoint(container, &corners[3]);
  if (nContains == 0)
  {
    return (unsigned int) 0;
  }
  else
    if (nContains < 4)
  {
    return (unsigned int) 1;
  }
  else
  {
    return (unsigned int) 2;
  }
}

float kmAABB2DiameterX(const kmAABB2 *aabb)
{
  return aabb->max.x - aabb->min.x;
}

float kmAABB2DiameterY(const kmAABB2 *aabb)
{
  return aabb->max.y - aabb->min.y;
}

kmVec2 *kmAABB2Centre(const kmAABB2 *aabb, kmVec2 *pOut)
{
  kmVec2Add(pOut, &aabb->min, &aabb->max);
  kmVec2Scale(pOut, pOut, 0.5);
  return pOut;
}

kmAABB2 *kmAABB2ExpandToContain(kmAABB2 *pOut, const kmAABB2 *pIn, const kmAABB2 *other)
{
  kmAABB2 result;
  result.min.x = (pIn->min.x < other->min.x) ? (pIn->min.x) : (other->min.x);
  result.max.x = (pIn->max.x > other->max.x) ? (pIn->max.x) : (other->max.x);
  result.min.y = (pIn->min.y < other->min.y) ? (pIn->min.y) : (other->min.y);
  result.max.y = (pIn->max.y > other->max.y) ? (pIn->max.y) : (other->max.y);
  kmAABB2Assign(pOut, &result);
  return pOut;
}

File: artifacts/heman/translate/src/aabb2.rs
use crate::*;

pub fn kmAABB2Sanitize<'a>(pOut: &'a mut KmAabb2, pIn: &KmAabb2) -> &'a mut KmAabb2 {
    // Handle x coordinates
    if pIn.min.x <= pIn.max.x {
        pOut.min.x = pIn.min.x;
        pOut.max.x = pIn.max.x;
    } else {
        pOut.min.x = pIn.max.x;
        pOut.max.x = pIn.min.x;
    }

    // Handle y coordinates
    if pIn.min.y <= pIn.max.y {
        pOut.min.y = pIn.min.y;
        pOut.max.y = pIn.max.y;
    } else {
        pOut.min.y = pIn.max.y;
        pOut.max.y = pIn.min.y;
    }

    pOut
}
pub fn kmAABB2DiameterX(aabb: &KmAabb2) -> f32 {
    aabb.max.x - aabb.min.x
}
pub fn kmAABB2DiameterY(aabb: &KmAabb2) -> f32 {
    aabb.max.y - aabb.min.y
}
pub fn kmAABB2ContainsPoint(pBox: &KmAabb2, pPoint: &KmVec2) -> bool {
    ((pPoint.x >= pBox.min.x) && (pPoint.x <= pBox.max.x)) && 
    ((pPoint.y >= pBox.min.y) && (pPoint.y <= pBox.max.y))
}
pub fn kmAABB2Assign<'a>(pOut: Option<&'a mut KmAabb2>, pIn: Option<&'a KmAabb2>) -> Option<&'a mut KmAabb2> {
    // Check for None (equivalent to NULL checks in C)
    if pOut.is_none() || pIn.is_none() {
        return None;
    }

    // Unwrap safely since we've checked for None
    let pOut = pOut.unwrap();
    let pIn = pIn.unwrap();

    // Assign min and max vectors
    kmVec2Assign(Some(&mut pOut.min), Some(&pIn.min));
    kmVec2Assign(Some(&mut pOut.max), Some(&pIn.max));

    Some(pOut)
}
pub fn kmAABB2Scale<'a>(pOut: &'a mut KmAabb2, pIn: &'a KmAabb2, s: f32) -> &'a mut KmAabb2 {
    {
        let max = &mut pOut.max;
        kmVec2Scale(max, &pIn.max, s);
    }
    {
        let min = &mut pOut.min;
        kmVec2Scale(min, &pIn.min, s);
    }
    pOut
}
pub fn kmAABB2ExpandToContain<'a>(
    pOut: Option<&'a mut KmAabb2>,
    pIn: Option<&'a KmAabb2>,
    other: Option<&'a KmAabb2>,
) -> Option<&'a mut KmAabb2> {
    // Check for None values (equivalent to NULL checks in C)
    if pIn.is_none() || other.is_none() {
        return Option::None;
    }

    // Unwrap safely since we've checked for None
    let pIn = pIn.unwrap();
    let other = other.unwrap();

    // Calculate the expanded AABB
    let result = KmAabb2 {
        min: KmVec2 {
            x: pIn.min.x.min(other.min.x),
            y: pIn.min.y.min(other.min.y),
        },
        max: KmVec2 {
            x: pIn.max.x.max(other.max.x),
            y: pIn.max.y.max(other.max.y),
        },
    };

    // Assign the result to pOut and return it
    if let Some(out) = pOut {
        *out = result.clone();
        Some(out)
    } else {
        Option::None
    }
}
pub fn kmAABB2Initialize<'a>(
    pBox: Option<&'a mut KmAabb2>,
    centre: Option<&KmVec2>,
    width: f32,
    height: f32,
    _depth: f32,  // Note: depth is unused in the original C code
) -> Option<&'a mut KmAabb2> {
    // Check for NULL pBox (translated to Option)
    let pBox = pBox?;

    // Handle centre pointer (translated to Option)
    let point = match centre {
        Some(centre) => KmVec2 {
            x: centre.x,
            y: centre.y,
        },
        None => {
            let mut origin = KmVec2 { x: 0.0, y: 0.0 };
            kmVec2Fill(&mut origin, 0.0, 0.0);
            origin
        }
    };

    // Calculate bounds
    pBox.min.x = point.x - (width / 2.0);
    pBox.min.y = point.y - (height / 2.0);
    pBox.max.x = point.x + (width / 2.0);
    pBox.max.y = point.y + (height / 2.0);

    Some(pBox)
}
pub fn kmAABB2Translate<'a>(
    pOut: &'a mut KmAabb2,
    pIn: &KmAabb2,
    translation: &KmVec2,
) -> &'a mut KmAabb2 {
    {
        let min_out = &mut pOut.min;
        kmVec2Add(min_out, &pIn.min, translation);
    }
    {
        let max_out = &mut pOut.max;
        kmVec2Add(max_out, &pIn.max, translation);
    }
    pOut
}
pub fn kmAABB2ScaleWithPivot<'a>(
    pOut: &'a mut KmAabb2,
    pIn: &'a KmAabb2,
    pivot: &KmVec2,
    s: f32,
) -> &'a mut KmAabb2 {
    let translate = KmVec2 {
        x: -pivot.x,
        y: -pivot.y,
    };
    
    // First translation
    let pOut = kmAABB2Translate(pOut, pIn, &translate);
    
    // Scaling
    let pOut = kmAABB2Scale(pOut, pIn, s);
    
    // Second translation
    kmAABB2Translate(pOut, pIn, pivot)
}
pub fn kmAABB2ContainsAABB(container: &KmAabb2, to_check: &KmAabb2) -> u32 {
    let mut corners = [
        KmVec2 { x: 0.0, y: 0.0 },
        KmVec2 { x: 0.0, y: 0.0 },
        KmVec2 { x: 0.0, y: 0.0 },
        KmVec2 { x: 0.0, y: 0.0 },
    ];
    
    {
        let corner = &mut corners[0];
        kmVec2Fill(corner, to_check.min.x, to_check.min.y);
    }
    {
        let corner = &mut corners[1];
        kmVec2Fill(corner, to_check.max.x, to_check.min.y);
    }
    {
        let corner = &mut corners[2];
        kmVec2Fill(corner, to_check.max.x, to_check.max.y);
    }
    {
        let corner = &mut corners[3];
        kmVec2Fill(corner, to_check.min.x, to_check.max.y);
    }

    let n_contains = kmAABB2ContainsPoint(container, &corners[0]) as u32
        + kmAABB2ContainsPoint(container, &corners[1]) as u32
        + kmAABB2ContainsPoint(container, &corners[2]) as u32
        + kmAABB2ContainsPoint(container, &corners[3]) as u32;

    match n_contains {
        0 => 0,
        1..=3 => 1,
        _ => 2,
    }
}
pub fn kmAABB2Centre<'a>(aabb: &KmAabb2, pOut: &'a mut KmVec2) -> &'a mut KmVec2 {
    let mut temp = KmVec2 { x: 0.0, y: 0.0 };
    kmVec2Add(&mut temp, &aabb.min, &aabb.max);
    kmVec2Scale(pOut, &temp, 0.5);
    pOut
}
